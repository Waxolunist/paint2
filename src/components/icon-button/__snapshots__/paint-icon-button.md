# `paint-icon-button`

#### `renders correctly`

```html
<div
  class="animated elevate elevate-2 elevated-3 icon-button icon-wrapper shadow"
>
  <slot> </slot>
</div>
```

#### `renders correctly when active and with slot`

```html
<div
  class="animated elevate elevate-2 elevated-3 icon-button icon-wrapper shadow"
>
  <slot> </slot>
</div>
```

```html
Foo
```

#### `renders correctly after pointerdown`

```html
<div
  class="animated clicked elevate elevate-2 elevated-3 icon-button icon-wrapper shadow"
>
  <slot> </slot>
</div>
```

#### `renders correctly after pointerdown and up`

```html
<div
  class="animated elevate elevate-2 elevated-3 icon-button icon-wrapper shadow"
>
  <slot> </slot>
</div>
```
