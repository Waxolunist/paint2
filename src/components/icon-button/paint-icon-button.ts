import {
  css,
  customElement,
  html,
  query,
  eventOptions,
  LitElement,
  CSSResult,
  TemplateResult,
} from 'lit-element';
import {ShadowStyles, AnimatedStyles} from '../../styles/shared-styles';
import {property} from 'lit-element';

@customElement('paint-icon-button')
export class IconButton extends LitElement {
  static get styles(): CSSResult[] {
    // language=CSS
    return [
      AnimatedStyles,
      ShadowStyles,
      css`
        ::slotted(svg) {
          width: calc(var(--icon-size) - 6px);
          height: calc(var(--icon-size) - 6px);
          padding-top: 3px;
          display: block;
          margin: auto;
        }

        .icon-wrapper {
          width: 100%;
          height: 100%;
          border-radius: var(--paint-icon-button-border-radius, 50%);
          background-color: var(--paint-icon-button-background-color, white);
          display: block;
          transition-property: box-shadow, border-radius !important;
          will-change: box-shadow, border-radius !important;
        }
      `,
    ];
  }

  @query('.icon-button')
  private button!: HTMLElement;

  @property({type: Boolean})
  active = false;

  render(): TemplateResult {
    return html`
      <div
        class="icon-button icon-wrapper shadow animated elevate elevate-2 elevated-3"
        @pointerdown="${this.pointerDown}"
        @pointerup="${this.pointerUp}"
        @pointercancel="${this.pointerUp}"
        @click="${this.clicked}"
      >
        <slot></slot>
      </div>
    `;
  }

  @eventOptions({capture: true, passive: true})
  private pointerDown(e: PointerEvent): void {
    e.stopPropagation();
    this.button.classList.add('clicked');
    if (process?.env?.NODE_ENV !== 'test')
      this.button.setPointerCapture(e.pointerId);
  }

  @eventOptions({capture: true, passive: true})
  private pointerUp(e: PointerEvent): void {
    e.stopPropagation();
    this.button.classList.remove('clicked');
    if (process?.env?.NODE_ENV !== 'test')
      this.button.releasePointerCapture(e.pointerId);
  }

  @eventOptions({capture: true, passive: false})
  private clicked(e: MouseEvent): void {
    e.preventDefault();
    e.stopPropagation();
    this.dispatchEvent(
      new CustomEvent('icon-clicked', {
        bubbles: true,
        composed: true,
      })
    );
  }
}
