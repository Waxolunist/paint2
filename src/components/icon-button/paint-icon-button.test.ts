import {fireClickEvent, firePointerEvent} from '../../test/pointerevents';
import {fixture, html, oneEvent} from '@open-wc/testing';
import {IconButton} from './paint-icon-button';

describe('paint-icon-button', () => {
  it('is defined', () => {
    const el = document.createElement('paint-icon-button');
    expect(el).toBeInstanceOf(IconButton);
  });

  it('renders correctly', async () => {
    const element = await fixture(
      html` <paint-icon-button></paint-icon-button>`
    );
    expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
  });

  it('renders correctly when active and with slot', async () => {
    const element = await fixture(
      html` <paint-icon-button active>Foo</paint-icon-button>`
    );
    expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
  });

  it('renders correctly after pointerdown', async () => {
    const element = await fixture(
      html` <paint-icon-button></paint-icon-button>`
    );
    await firePointerEvent(element, ['down']);
    expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
  });

  it('renders correctly after pointerdown and up', async () => {
    const element = await fixture(
      html` <paint-icon-button></paint-icon-button>`
    );
    await firePointerEvent(element, ['down', 'up']);
    expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
  });

  it('fires event after pointer up', async () => {
    const el = await fixture(html` <paint-icon-button></paint-icon-button>`);
    fireClickEvent(el);
    const {detail} = await oneEvent(el, 'icon-clicked');
    expect(detail).not.toBeUndefined();
  });
});
