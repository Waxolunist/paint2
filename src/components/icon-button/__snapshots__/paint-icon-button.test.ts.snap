// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paint-icon-button renders correctly 1`] = `
"
      <div class=\\"icon-button icon-wrapper shadow animated elevate elevate-2 elevated-3\\">
        <slot></slot>
      </div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
  .shadow.animated.elevate {
    transition-property: box-shadow;
    box-shadow: none;
    will-change: box-shadow;
  }

  .shadow.animated.elevate.clicked {
    box-shadow: none !important;
    background-color: #ddd;
  }

  .shadow.elevate.elevate-1,
  .shadow.elevate.elevated-1:hover {
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
      0 3px 1px -2px rgba(0, 0, 0, 0.2);
  }

  .shadow.elevate.elevate-2,
  .shadow.elevate.elevated-2:hover {
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14),
      0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.4);
  }

  .shadow.elevate.elevate-3,
  .shadow.elevate.elevated-3:hover {
    box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14),
      0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.4);
  }
</style><style>
      ::slotted(svg) {
        width: calc(var(--icon-size) - 6px);
        height: calc(var(--icon-size) - 6px);
        padding-top: 3px;
        display: block;
        margin: auto;
      }

      .icon-wrapper {
        width: 100%;
        height: 100%;
        border-radius: var(--paint-icon-button-border-radius, 50%);
        background-color: var(--paint-icon-button-background-color, white);
        display: block;
        transition-property: box-shadow, border-radius !important;
        will-change: box-shadow, border-radius !important;
      }
    </style>"
`;

exports[`paint-icon-button renders correctly after pointerdown 1`] = `
"
      <div class=\\"icon-button icon-wrapper shadow animated elevate elevate-2 elevated-3 clicked\\">
        <slot></slot>
      </div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
  .shadow.animated.elevate {
    transition-property: box-shadow;
    box-shadow: none;
    will-change: box-shadow;
  }

  .shadow.animated.elevate.clicked {
    box-shadow: none !important;
    background-color: #ddd;
  }

  .shadow.elevate.elevate-1,
  .shadow.elevate.elevated-1:hover {
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
      0 3px 1px -2px rgba(0, 0, 0, 0.2);
  }

  .shadow.elevate.elevate-2,
  .shadow.elevate.elevated-2:hover {
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14),
      0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.4);
  }

  .shadow.elevate.elevate-3,
  .shadow.elevate.elevated-3:hover {
    box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14),
      0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.4);
  }
</style><style>
      ::slotted(svg) {
        width: calc(var(--icon-size) - 6px);
        height: calc(var(--icon-size) - 6px);
        padding-top: 3px;
        display: block;
        margin: auto;
      }

      .icon-wrapper {
        width: 100%;
        height: 100%;
        border-radius: var(--paint-icon-button-border-radius, 50%);
        background-color: var(--paint-icon-button-background-color, white);
        display: block;
        transition-property: box-shadow, border-radius !important;
        will-change: box-shadow, border-radius !important;
      }
    </style>"
`;

exports[`paint-icon-button renders correctly after pointerdown and up 1`] = `
"
      <div class=\\"icon-button icon-wrapper shadow animated elevate elevate-2 elevated-3\\">
        <slot></slot>
      </div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
  .shadow.animated.elevate {
    transition-property: box-shadow;
    box-shadow: none;
    will-change: box-shadow;
  }

  .shadow.animated.elevate.clicked {
    box-shadow: none !important;
    background-color: #ddd;
  }

  .shadow.elevate.elevate-1,
  .shadow.elevate.elevated-1:hover {
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
      0 3px 1px -2px rgba(0, 0, 0, 0.2);
  }

  .shadow.elevate.elevate-2,
  .shadow.elevate.elevated-2:hover {
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14),
      0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.4);
  }

  .shadow.elevate.elevate-3,
  .shadow.elevate.elevated-3:hover {
    box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14),
      0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.4);
  }
</style><style>
      ::slotted(svg) {
        width: calc(var(--icon-size) - 6px);
        height: calc(var(--icon-size) - 6px);
        padding-top: 3px;
        display: block;
        margin: auto;
      }

      .icon-wrapper {
        width: 100%;
        height: 100%;
        border-radius: var(--paint-icon-button-border-radius, 50%);
        background-color: var(--paint-icon-button-background-color, white);
        display: block;
        transition-property: box-shadow, border-radius !important;
        will-change: box-shadow, border-radius !important;
      }
    </style>"
`;

exports[`paint-icon-button renders correctly when active and with slot 1`] = `
"
      <div class=\\"icon-button icon-wrapper shadow animated elevate elevate-2 elevated-3\\">
        <slot></slot>
      </div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
  .shadow.animated.elevate {
    transition-property: box-shadow;
    box-shadow: none;
    will-change: box-shadow;
  }

  .shadow.animated.elevate.clicked {
    box-shadow: none !important;
    background-color: #ddd;
  }

  .shadow.elevate.elevate-1,
  .shadow.elevate.elevated-1:hover {
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
      0 3px 1px -2px rgba(0, 0, 0, 0.2);
  }

  .shadow.elevate.elevate-2,
  .shadow.elevate.elevated-2:hover {
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14),
      0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.4);
  }

  .shadow.elevate.elevate-3,
  .shadow.elevate.elevated-3:hover {
    box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14),
      0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.4);
  }
</style><style>
      ::slotted(svg) {
        width: calc(var(--icon-size) - 6px);
        height: calc(var(--icon-size) - 6px);
        padding-top: 3px;
        display: block;
        margin: auto;
      }

      .icon-wrapper {
        width: 100%;
        height: 100%;
        border-radius: var(--paint-icon-button-border-radius, 50%);
        background-color: var(--paint-icon-button-background-color, white);
        display: block;
        transition-property: box-shadow, border-radius !important;
        will-change: box-shadow, border-radius !important;
      }
    </style>"
`;
